File: 2020/2020W42/preppindataw42.R
library(dplyr)
library(tidyr)
library(readxl)
library(lubridate)

period <- data.frame(`Today` = as.Date('2020-10-09',format='%Y-%m-%d')) %>%
  mutate('Week' = as.integer(strftime(`Today`,format = '%U')),
         'Days' = as.integer(strftime(`Today`,format = '%w'))) %>%
  merge(data.frame(`Offset` = seq(0, -1))) %>%
  mutate('Category' = if_else(`Offset`==0,'This Year','Last Year'),
         'Year' = as.integer(strftime(`Today`,format = '%Y'))+`Offset`,
         'End Date' = as.Date(paste(`Year`,`Week`,`Days`),format='%Y %U %w'),
         'WTD' = floor_date(`End Date`, 'weeks'),
         'MTD' = floor_date(`End Date`, 'months'),
         'YTD' = floor_date(`End Date`, 'years')) %>%
  select(-c('Today','Week','Days','Offset')) %>%
  pivot_longer(.,cols=ends_with('TD'), names_to='Time Period', values_to='Start Date')

txn <- read_excel('F:/Data/Transactions.xlsx') %>%
  filter(`TransactionDate` >= period %>% summarise('min'= min(`Start Date`)) %>% .[[1]]) %>%
  mutate('From Date' = as.Date(`TransactionDate`),
         'To Date' = as.Date(`TransactionDate`),
         'Year' = as.integer(strftime(`TransactionDate`,format = '%Y')),
         'Source' = 'Transaction') %>%
  select(c('Year', 'From Date', 'To Date', 'ProductName', 'Quantity', 'Income', 'Source'))

target <- read_excel('F:/Data/Targets.xlsx') %>%
  rename('Quantity' = 'Quantity Target',
         'Income' = 'Income Target') %>%
  mutate('From Date'= floor_date(as.Date(paste0(`Year`,'-01-01'),format='%Y-%m-%d'),'weeks')+(`Week`-1)*7,
         'To Date' = `From Date`+6,
         'Source' = 'Target') %>%
  select(-'Week')

final <- rbind(txn, target) %>%
  merge(., period, by='Year') %>%
  filter((`From Date`>=`Start Date` & `To Date`<=`End Date`) | (`From Date`<=`End Date` & `To Date`>=`Start Date`)) %>%
  rowwise() %>%
  mutate('Category' = if_else(`Source`=='Transaction',`Category`,`Source`),
         'Modifier' = as.integer(difftime(min(`End Date`,`To Date`),max(`Start Date`,`From Date`), units='day'))+1) %>%
  ungroup() %>%
  select(c('ProductName', 'Category', 'Time Period', 'Modifier', 'Quantity', 'Income')) %>%
  pivot_longer(.,cols=c('Quantity', 'Income'), names_to='Metric', values_to='Value') %>%
  group_by(`ProductName`,`Category`, `Time Period`, `Metric`) %>%
  summarise('Value' = round(sum(if_else(`Category`=='Target',`Value`/7*`Modifier`,`Value`)))) %>%
  pivot_wider(., names_from='Category', values_from='Value', values_fn=list(`Value`=sum)) %>%
  mutate('% Differece to Last Year' = round(((`This Year`-`Last Year`)/`Last Year`),2),
         '% Differece to Target' = round(((`This Year`-`Target`)/`Target`),2)) %>%
  select(c('ProductName', 'Metric', 'Time Period', 'This Year', 'Last Year', 'Target', '% Differece to Last Year', '% Differece to Target'))
  
View(final)

File: 2020/2020W42/PreppindataW42.py
import pandas as pd
import datetime 
from datetime import datetime as dt,timedelta

today = datetime.date(2020, 10, 9)
period = pd.DataFrame({'Category':['This Year','Last Year'], 
                       'End Date':[today, dt.strptime(str(int(dt.strftime(today, "%Y"))-1)+dt.strftime(today, "-%U-%w"), "%Y-%U-%w").date()]})
period['WTD']  = period['End Date'].apply(lambda x: x-timedelta(days=int(dt.strftime(x, "%w"))))
period['MTD']  = period['End Date'].apply(lambda x: x.replace(day=1))
period['YTD']  = period['End Date'].apply(lambda x: x.replace(day=1,month=1))
period = period.melt(id_vars = ['Category', 'End Date'], value_name = 'Start Date', var_name='Time Period')
period['Year'] = period['End Date'].apply(lambda x: int(dt.strftime(x, "%Y")))

txn = pd.read_excel(pd.ExcelFile("F:/Data/Transactions.xlsx"),sheet_names=1)  
target = pd.read_excel(pd.ExcelFile("F:/Data/Targets.xlsx"),sheet_names=1)

txn['From Date'] = txn['TransactionDate'].dt.date 
txn['To Date'] = txn['TransactionDate'].dt.date 
txn = txn[txn['TransactionDate'].dt.date>=period['Start Date'].min()][['From Date', 'To Date', 'ProductName', 'Quantity', 'Income']].copy()
txn['Year'] = txn['To Date'].apply(lambda x: int(dt.strftime(x, "%Y")))

target = target.rename(columns={'Quantity Target':'Quantity','Income Target':'Income'})
target['From Date'] = target.apply(lambda x: dt.strptime(f"{x['Year']}-{x['Week']-1}-0",'%Y-%U-%w').date(), axis=1)
target['To Date'] = target['From Date'].apply(lambda x: x+timedelta(days=6))
target = target[['Year', 'From Date', 'To Date', 'ProductName', 'Quantity', 'Income']].copy()
target['Source'] = 'Target'

final = pd.concat([txn, target], sort=False, ignore_index=True)
final = final.merge(period, on='Year', how='inner')
final['Category'] = final.apply(lambda x: x['Category'] if pd.isnull(x['Source']) else x['Source'], axis=1)
final = final[((final['From Date']>=final['Start Date']) & (final['To Date']<=final['End Date']))
              | ((final['From Date']<=final['End Date']) & (final['To Date']>=final['Start Date']))].copy()
final['Modifier'] = final.apply(lambda x: (min(x['End Date'],x['To Date'])-max(x['Start Date'],x['From Date'])).days+1, axis=1)
final = final[['ProductName', 'Category', 'Time Period', 'Modifier', 'Quantity', 'Income']].copy()
final = final.melt(id_vars = ['ProductName', 'Category', 'Time Period', 'Modifier'], value_name = 'Value', var_name='Metric')
final['Value'] = final.apply(lambda x: x['Value']/7*x['Modifier'] if x['Category']=='Target' else x['Value'], axis=1)
final = final.groupby(['ProductName','Category', 'Time Period', 'Metric'], as_index=False).agg({'Value':'sum'})
final['Value'] = final['Value'].apply(lambda x: round(x))
final = final.pivot_table(index=['ProductName', 'Metric', 'Time Period'], columns='Category', values='Value', aggfunc=sum).reset_index()
final['% Differece to Last Year'] = round((final['This Year']-final['Last Year'])/final['Last Year'],2)
final['% Differece to Target'] = round((final['This Year']-final['Target'])/final['Target'],2)
final = final[['ProductName', 'Metric', 'Time Period', 'This Year', 'Last Year', 'Target', '% Differece to Last Year', '% Differece to Target']]
